File: rest/crdp/python/CRDPBulkOperation.py
import json
import ssl
import sys
import requests
from urllib3 import disable_warnings, exceptions
from requests.auth import HTTPBasicAuth

BATCH_LIMIT = 5000

def main(args):
    crdp_container_ip = "yourip"
    batch_size = 250
    protect_url = f"http://{crdp_container_ip}:8090/v1/protect"
    debug = False
    protection_profile = "plain-nbr-internal"
    mode_of_operation = "regular"
    
    if args[1] != "null":
        debug = args[1].lower() == "true"
        
    if args[2] != "null":
        mode_of_operation = args[2]
        
    if mode_of_operation.lower() == "bulk":
        protect_url = protect_url + "bulk"
        
    file_path = r"C:\data\emp-id1k.txt"
    total_records = 0
    
    if mode_of_operation.lower() == "regular":
        total_records = protect_from_file(protect_url, debug, file_path, protection_profile)
        nbrofchunks = 1
    else:
        strings = read_strings_from_file(file_path)
        total_number_of_records = len(strings)
        index = 0
        nbrofchunks = 0
        records_left = total_number_of_records
        
        batch_nnn = []
        if batch_size > total_number_of_records:
            batch_size = total_number_of_records
        if batch_size >= BATCH_LIMIT:
            batch_size = BATCH_LIMIT
            
        while index < total_number_of_records:
            for i in range(batch_size):
                if index < total_number_of_records:
                    batch_nnn.append(strings[index])
                    index += 1
                    records_left -= 1
            
            total_records += protect_from_file_bulk(protect_url, debug, batch_nnn, batch_size, index, total_number_of_records, records_left, protection_profile)
            batch_nnn = []
            nbrofchunks += 1

    print(f"Total records = {total_records}")

def protect_from_file_bulk(url, debug, strings, batch_size, index, total_number_of_records, records_left, protection_profile):
    total_nb_of_records = 0
    count = 0
    crdp_payload_array = []
    
    for str_ in strings:
        total_nb_of_records += 1
        crdp_payload_array.append(str_)
        count += 1
        
        if count == batch_size:
            crdp_payload = {"data_array": crdp_payload_array, "protection_policy_name": protection_profile}
            json_body = json.dumps(crdp_payload)
            if debug:
                print("json body:", json_body)
            make_cm_call(json_body, url, debug)
            crdp_payload_array = []
            count = 0
            
    if count > 0:
        crdp_payload = {"data_array": crdp_payload_array, "protection_policy_name": protection_profile}
        json_body = json.dumps(crdp_payload)
        if debug:
            print("json body:", json_body)
        make_cm_call(json_body, url, debug)

    return total_nb_of_records

def protect_from_file(url, debug, file_path, protection_profile):
    total_nb_of_records = 0
    
    with open(file_path, 'r') as file:
        for line in file:
            total_nb_of_records += 1
            crdp_payload = {
                "protection_policy_name": protection_profile,
                "data": line.strip()
            }
            json_body = json.dumps(crdp_payload)
            if debug:
                print("jsonBody", json_body)
            make_cm_call(json_body, url, debug)
    
    return total_nb_of_records

def disable_cert_validation():
    disable_warnings(exceptions.InsecureRequestWarning)
    ssl._create_default_https_context = ssl._create_unverified_context

def make_cm_call(payload, url, debug):
    disable_cert_validation()
    
    headers = {'Content-Type': 'application/json'}
    
    response = requests.post(url, data=payload, headers=headers, verify=False)
    
    if debug:
        print(f"Response Code: {response.status_code}")
        print(f"Response Body: {response.text}")
    
    return response.status_code

def read_strings_from_file(file_path):
    with open(file_path, 'r') as file:
        return [line.strip() for line in file]

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print( len(sys.argv))
        print("Usage: script.py <debug> <mode_of_operation>")
        sys.exit(1)
    
    # sys.argv[0] is the script name, so pass args starting from index 1
    args = sys.argv[1:]
    print(f"Debug mode: {args[1]}, Mode of operation: {args[2]}")
    main(args)

File: rest/crdp/java/src/main/java/com/example/CRDPBulkOperation.java
package com.example;

/**
* Sample code is provided for educational purposes.
* No warranty of any kind, either expressed or implied by fact or law.
* Use of this item is not restricted by copyright or license terms.
*/
// Standard JCE classes.
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

/**
 * This sample shows how to use the protect/reveal REST API using both the regular protect and the protectbulk.
 * 
 */
public class CRDPBulkOperation {

	private static int BATCHLIMIT = 5000;

	public static void main(String[] args) throws Exception {

		String crdp_container_IP = "192.168.159.143";
		int batchsize = 2;
		String protectURL = "http://" + crdp_container_IP + ":8090/v1/protect";
		boolean debug = false;

		String protection_profile = "plain-nbr-internal";
		String modeofoperation = "regular";

		if (!args[0].contains("null")) {
			debug = args[0].equalsIgnoreCase("true");
		}

		if (!args[1].contains("null")) {
			modeofoperation = args[1];
		}
		if (modeofoperation.equalsIgnoreCase("bulk"))
			protectURL = protectURL + "bulk";

		int index = 0;
		int totalrecords = 0;
		int recordcnter = 0;
		long startTime = System.currentTimeMillis();
		int nbrofchunks = 0;

		//emp-id1k.txt, emp-id5k.txt,emp-nbr-smallout.txt
		String filePath = "C:\\data\\emp-nbr-smallout.txt"; // Change



		if (modeofoperation.equalsIgnoreCase("regular")) {
			totalrecords = protectFromFile(protectURL, debug, filePath, protection_profile);
			nbrofchunks = 1;
		} else {
			List<String> strings = readStringsFromFile(filePath);
			int totalnumberofrecords = strings.size();
			int recordsleft = totalnumberofrecords;

			List<String> batchnnn = new ArrayList<>();
			startTime = System.currentTimeMillis();

			if (batchsize > totalnumberofrecords)
				batchsize = totalnumberofrecords;
			if (batchsize >= BATCHLIMIT)
				batchsize = BATCHLIMIT;

			while (index < totalnumberofrecords) {

				for (int i = 0; i < batchsize && index < totalnumberofrecords; i++) {
					batchnnn.add(strings.get(recordcnter));
					recordcnter++;
					index++;
					recordsleft--;

				}

				totalrecords = totalrecords + protectFromFileBulk(protectURL, debug, batchnnn, batchsize, index,
						totalnumberofrecords, recordsleft, protection_profile);
				batchnnn = new ArrayList<>();
				nbrofchunks++;

			}
		}

		long endTime = System.currentTimeMillis();
		long totalTime = endTime - startTime;
		System.out.println("Time taken CDRP test: " + totalTime + " milliseconds");
		if (modeofoperation.equalsIgnoreCase("bulk")) {
			System.out.println("BatchSize: " + batchsize);
			System.out.println("Nbrofchunks : " + nbrofchunks);
		}
		System.out.println("Total records = " + totalrecords);

	}

	public static int protectFromFileBulk(String url, boolean debug, List<String> strings, int batchsize, int index,
			int totalnumberofrecords, int recordsleft, String protection_profile) throws Exception {

		JsonObject crdp_payload = new JsonObject();
		String jsonBody = null;
		JsonArray crdp_payload_array = new JsonArray();
		String inputdataarray = null;
		StringBuffer protection_policy_buff = new StringBuffer();

		int totalnbrofrecords = 0;

		String line = "";
		int count = 0;
		int stopper = batchsize;
		if (index == totalnumberofrecords)
			stopper = recordsleft + 1;

		String formattedElement = String.format("\"protection_policy_name\" : \"%s\"", protection_profile);
		protection_policy_buff.append(formattedElement);
		protection_policy_buff.append(",");

		// Loop is only needed if you have different policies for each data value in the array.
		// Example:
		/*
		 * {"protection_policy_name": "plain-nbr-internal","protection_policy_name": "plain-alpha-internal",
		 * "data_array": [ "1234567812345678", "abcdef" ]}
		 * 
		 * for (int i = 0; i < stopper; i++) {
		 * 
		 * String formattedElement = String.format("\"protection_policy_name\" : \"%s\"", protection_profile);
		 * protection_policy_buff.append(formattedElement); protection_policy_buff.append(",");
		 * 
		 * }
		 */

		for (String str : strings) {
			totalnbrofrecords++;

			crdp_payload_array.add(str);
			count++;

			if (count == batchsize) {

				crdp_payload.add("data_array", crdp_payload_array);
				inputdataarray = crdp_payload.toString();
				protection_policy_buff.append(inputdataarray);
				inputdataarray = protection_policy_buff.toString();
				jsonBody = inputdataarray.replace("{", " ");

				jsonBody = "{" + jsonBody;

				if (debug)
					System.out.println("json body ---------------" + jsonBody);

				makeCMCall(jsonBody, url, debug);

				formattedElement = String.format("\"protection_policy_name\" : \"%s\"", protection_profile);
				protection_policy_buff.append(formattedElement);
				protection_policy_buff.append(",");
				// Loop is only needed if you have different policies for each data value in the array.
				/*
				 * for (int i = 0; i < stopper; i++) { String formattedElement =
				 * String.format("\"protection_policy_name\" : \"%s\"", protection_profile);
				 * protection_policy_buff.append(formattedElement); protection_policy_buff.append(",");
				 * 
				 * }
				 */
				count = 0;
			}
			line = str;
		}

		if (count > 0) {
			crdp_payload.add("data_array", crdp_payload_array);
			inputdataarray = crdp_payload.toString();
			protection_policy_buff.append(inputdataarray);
			inputdataarray = protection_policy_buff.toString();
			jsonBody = inputdataarray.replace("{", " ");

			jsonBody = "{" + jsonBody;

			if (debug) {
				System.out.println(jsonBody.toString());
				System.out.println("count =" + count);
			}
			makeCMCall(jsonBody, url, debug);

		}

		return totalnbrofrecords;

	}

	public static int protectFromFile(String url, boolean debug, String filePath, String protection_profile)
			throws Exception {

		JsonObject crdp_payload = new JsonObject();
		String jsonBody = null;
		String inputdataarray = null;

		int totalnbrofrecords = 0;
		try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
			String line;

			crdp_payload.addProperty("protection_policy_name", protection_profile);

			while ((line = br.readLine()) != null) {
				totalnbrofrecords++;
				crdp_payload.addProperty("data", line);
				inputdataarray = crdp_payload.toString();
				jsonBody = inputdataarray.replace("{", " ");
				jsonBody = "{" + jsonBody;
				if (debug)
					System.out.println("jsonBody        " + jsonBody);
				makeCMCall(jsonBody, url, debug);
				crdp_payload = new JsonObject();
				crdp_payload.addProperty("protection_policy_name", protection_profile);
			}

		} catch (IOException e) {
			e.printStackTrace();
		}

		return totalnbrofrecords;

	}

	public static void disableCertValidation() throws Exception {
		TrustManager[] trustAllCerts = new TrustManager[] { new X509TrustManager() {
			public java.security.cert.X509Certificate[] getAcceptedIssuers() {
				return null;
			}

			public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType) {
			}

			public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {
			}
		} };

		// Install the all-trusting trust manager
		try {
			SSLContext sc = SSLContext.getInstance("SSL");
			sc.init(null, trustAllCerts, new java.security.SecureRandom());
			HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
		} catch (Exception e) {
			e.printStackTrace();
		}

		// Create all-trusting host name verifier
		HostnameVerifier allHostsValid = new HostnameVerifier() {
			public boolean verify(String hostname, SSLSession session) {
				return true;
			}
		};

		// Install the all-trusting host verifier
		HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
	}

	public static int makeCMCall(String payload, String url, boolean debug) throws Exception {

		disableCertValidation();

		// Create URL object
		URL apiUrl = new URL(url);
		// Create connection object
		HttpURLConnection connection = (HttpURLConnection) apiUrl.openConnection();
		connection.setRequestMethod("POST");
		connection.setRequestProperty("Content-Type", "application/json");
		connection.setDoOutput(true);
		// connection.setRequestProperty("Authorization", newtoken);
		// Send request
		OutputStream outputStream = connection.getOutputStream();
		outputStream.write(payload.getBytes());
		outputStream.flush();
		outputStream.close();

		// Get response
		int responseCode = connection.getResponseCode();
		BufferedReader reader;
		if (responseCode >= 200 && responseCode < 300) {
			reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
		} else {
			reader = new BufferedReader(new InputStreamReader(connection.getErrorStream()));
		}

		// Read response
		String lineresponse;
		StringBuilder response = new StringBuilder();
		while ((lineresponse = reader.readLine()) != null) {
			response.append(lineresponse);
		}
		reader.close();

		// Print response
		if (debug) {
			System.out.println("Response Code: " + responseCode);
			System.out.println("Response Body: " + response.toString());
		}
		// Disconnect connection
		connection.disconnect();

		return responseCode;

	}

	public static List<String> readStringsFromFile(String filePath) {
		List<String> strings = new ArrayList<>();

		try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
			String line;
			while ((line = reader.readLine()) != null) {
				strings.add(line);
			}
		} catch (IOException e) {
			System.err.println("Error reading file: " + e.getMessage());
		}

		return strings;
	}

}
